### ISSUE TITLE
noUnusedLocals doesn't detect recursive closure functions

### ISSUE THREAD

**Issue by falsandtru** [Microsoft/TypeScript#18904]
related: #16078

cc @andy-ms 

**TypeScript Version:**  master

**Code**

```ts
export {}

function f() {
  f;
  function g() {
    g;
  }
}
```

**Expected behavior:**

```
$ node built/local/tsc.js index.ts --noEmit --noUnusedLocals
index.ts(3,10): error TS6133: 'f' is declared but its value is never read.
index.ts(5,12): error TS6133: 'g' is declared but its value is never read.
```

**Actual behavior:**

```
$ node built/local/tsc.js index.ts --noEmit --noUnusedLocals
index.ts(3,10): error TS6133: 'f' is declared but its value is never read.
```

**Comment by RyanCavanaugh** [comment]
Need more ☕️
